### PR TITLE
Wrap keycloak-connect

### PR DESCRIPTION
**Pull request by ashtonmeuser**
Not to be merged just yet. This is merely a preview of the wrapper around the [Keycloak connect library](https://www.npmjs.com/package/keycloak-connect). Simplifies getting user info (name, roles, region, etc.) using the connect middleware `getUserInfo()` method. Also saves realm and client info so we don't have to pull from `process.env` all the time.

### CODE REVIEW

File: server/keycloak.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      if (roles.length === 0) return true; // Allows any role
+      if (token.hasRole('superuser')) return true;
+      if (!roles.some((role) => token.hasRole(role))) return false;
+      if (token.isExpired()) return false;
+      return true;
+    };
+    return this.keycloakConnect.protect(allowRoles);
+  }
+
+  getUserInfo() { // Connect middleware for adding HCAP user info to request object
```
Comment (mauriciofreshw) [inline server/keycloak.js:61]: We should call this `getUserInfoMiddleware()` since it's returning a middleware and not the user info per se.

Comment (ashtonmeuser) [reply server/keycloak.js:61]: Will do!

File: openshift/keycloak.prep.yml [hunk 2]
```diff
@@ -34,6 +42,9 @@ objects:
     stringData:
       realm: ${KEYCLOAK_REALM}
       auth-url: ${KEYCLOAK_AUTH_URL}
-      fe-client-id: ${APP_NAME}${KEYCLOAK_FE_CLIENTID_SUFFIX}
+      fe-client-name: ${APP_NAME}${KEYCLOAK_FE_CLIENTID_SUFFIX}
```
Comment (fwkendall) [inline openshift/keycloak.prep.yml:45]: doesn't this name need to be changed in server.dc.yml as well?

File: server/keycloak.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    };
+  }
+
+  expressMiddleware(...args) { // Default connect middleware for Keycloak connect library
+    return this.keycloakConnect.middleware(...args);
+  }
+
+  allowRolesMiddleware(...roles) { // Connect middleware for limiting roles
+    const allowRoles = () => (token) => {
+      if (roles.length === 0) return true; // Allows any role
```
Comment (fwkendall) [inline server/keycloak.js:53]: is this temporary?

**Review by mauriciofreshw — APPROVED** [review]

File: server/keycloak.js (hunk 3 above)
Comment (ashtonmeuser) [reply server/keycloak.js:53]: This was to match the way that we were using `keycloak.protect()`. If no roles are provided, it just makes sure that the user is logged in. We could change this and all endpoints that depend on it. Looks like it's only the `/user` API endpoint that uses it.

File: openshift/keycloak.prep.yml (hunk 2 above)
Comment (ashtonmeuser) [reply openshift/keycloak.prep.yml]: Good catch. Fixed.

**Review by mauriciofreshw — APPROVED** [review]

File: server/server.js [hunk 4]
```diff
@@ -112,29 +87,26 @@ app.post(`${apiBaseUrl}/employees`,
...
-    }
-
+    if (req.fileError) return res.json({ status: 'Error', message: req.fileError });
     const response = await parseAndSaveParticipants(req.file);
     return res.json(response);
   }));
 
+// Get pending users from Keycloak
+app.get(`${apiBaseUrl}/pending-users`,
+  keycloak.allowRolesMiddleware('health_authority', 'ministry_of_health'),
```
Comment (fwkendall) [inline server/server.js:97]: should be just MOH here actually, HA approval was descoped

Comment (ashtonmeuser) [reply server/server.js]: Fixed.

**Review by fwkendall — DISMISSED** [review]

**Review by mauriciofreshw — APPROVED** [review]

File: server/keycloak.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    this.access_token = response.data.access_token;
+  }
+
+  async authenticateIfNeeded() {
+    // Race condition if token expires between this call and the desired authenticated call
+    const config = { headers: { Authorization: `Bearer ${this.access_token}` } };
+    try {
+      await axios.get(`${this.authUrl}/realms/${this.realm}/protocol/openid-connect/userinfo`, config);
+    } catch (error) {
+      await this.authenticateServiceAccount();
```
Comment (fwpushan) [inline server/keycloak.js:96]: As a suggestion, I recommend, do a little bit digging on error before re-authenticate, like check is it timeout / network issue / 401.

File: server/keycloak.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+const defaults = {
+  'ssl-required': 'external',
+  'confidential-port': 0,
+  'verify-token-audience': true,
+  'use-resource-role-mappings': true,
+  'policy-enforcer': {},
+};
+
+class Keycloak { // Wrapper class around keycloak-connect
```
Comment (fwpushan) [inline server/keycloak.js:21]: This is nice to have Keycloak manager like this. It helps us to Mock this class for testing.

**Review by fwkendall — APPROVED** [review]

**Review by mauriciofreshw — APPROVED** [review]

**Review by fwpushan — APPROVED** [review]

**Pull request merged**